File: bundles/ch.hsr.ifs.testframework/src/ch/hsr/ifs/testframework/launch/CustomisedLaunchConfigTab.java
/*******************************************************************************
 * Copyright (c) 2007-2011, IFS Institute for Software, HSR Rapperswil,
 * Switzerland, http://ifs.hsr.ch
 *
 * Permission to use, copy, and/or distribute this software for any
 * purpose without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 ******************************************************************************/
package ch.hsr.ifs.testframework.launch;

import org.eclipse.cdt.launch.ui.CLaunchConfigurationTab;
import org.eclipse.core.resources.IContainer;
import org.eclipse.core.resources.ResourcesPlugin;
import org.eclipse.core.runtime.CoreException;
import org.eclipse.core.runtime.IPath;
import org.eclipse.core.runtime.Path;
import org.eclipse.debug.core.ILaunchConfiguration;
import org.eclipse.debug.core.ILaunchConfigurationWorkingCopy;
import org.eclipse.debug.internal.ui.DebugUIPlugin;
import org.eclipse.debug.internal.ui.SWTFactory;
import org.eclipse.debug.ui.IDebugUIConstants;
import org.eclipse.jface.dialogs.IDialogSettings;
import org.eclipse.swt.SWT;
import org.eclipse.swt.events.ModifyListener;
import org.eclipse.swt.events.SelectionAdapter;
import org.eclipse.swt.events.SelectionEvent;
import org.eclipse.swt.graphics.Font;
import org.eclipse.swt.layout.GridData;
import org.eclipse.swt.layout.GridLayout;
import org.eclipse.swt.widgets.Button;
import org.eclipse.swt.widgets.Composite;
import org.eclipse.swt.widgets.Group;
import org.eclipse.swt.widgets.Label;
import org.eclipse.swt.widgets.Shell;
import org.eclipse.swt.widgets.Text;
import org.eclipse.ui.dialogs.ContainerSelectionDialog;

import ch.hsr.ifs.testframework.TestFrameworkPlugin;
import ch.hsr.ifs.testframework.model.Model;


/**
 * @since 3.0
 */
@SuppressWarnings("restriction")
public class CustomisedLaunchConfigTab extends CLaunchConfigurationTab {

   private static final String EMPTY_STRING        = "";
   public static final String  CUSTOM_SRC_PATH     = "customSrcPath";
   public static final String  USE_CUSTOM_SRC_PATH = "useCustomSrcPath";
   protected Label             descriptionLabel;
   private Button              fLocalRadioButton;
   private Button              fCustomSrcRadioButton;
   private Text                fCustomSrcLocationText;
   private Button              fCustomSrcLocationButton;

   protected Model cm = TestFrameworkPlugin.getModel();

   @Override
   public void createControl(Composite parent) {
      Font font = parent.getFont();
      Composite comp = new Composite(parent, SWT.NONE);
      createVerticalSpacer(comp, 1);

      GridLayout layout = new GridLayout(1, true);
      layout.numColumns = 2;
      layout.marginHeight = 0;
      layout.marginWidth = 5;
      comp.setLayout(layout);
      comp.setFont(font);
      GridData gd;

      String text = LaunchConfigurationsMessages.getString("CustomisedLaunchConfigTab.SourceFolderSelection");
      Group group = SWTFactory.createGroup(comp, text, 3, 2, GridData.FILL_HORIZONTAL);
      Composite n_comp = SWTFactory.createComposite(group, parent.getFont(), 3, 3, GridData.FILL_BOTH, 0, 0);
      descriptionLabel = new Label(n_comp, SWT.NONE);
      descriptionLabel.setText(LaunchConfigurationsMessages.getString("CustomisedLaunchConfigTab.SourceDescText"));
      gd = new GridData();
      gd.horizontalSpan = 3;
      descriptionLabel.setLayoutData(gd);

      fLocalRadioButton = createRadioButton(n_comp, LaunchConfigurationsMessages.getString("CustomisedLaunchConfigTab.Default"));
      gd = new GridData();
      gd.horizontalSpan = 3;
      fLocalRadioButton.setLayoutData(gd);

      fCustomSrcRadioButton = createRadioButton(n_comp, LaunchConfigurationsMessages.getString("CustomisedLaunchConfigTab.CustomPath"));
      fCustomSrcRadioButton.addSelectionListener(new SelectionAdapter() {

         @Override
         public void widgetSelected(SelectionEvent evt) {
            handleSharedRadioButtonSelected();
         }
      });
      fCustomSrcLocationText = SWTFactory.createSingleText(n_comp, 1);
      fCustomSrcLocationText.addModifyListener(fBasicModifyListener);
      fCustomSrcLocationButton = createPushButton(n_comp, EMPTY_STRING, null);
      fCustomSrcLocationButton.addSelectionListener(new SelectionAdapter() {

         @Override
         public void widgetSelected(SelectionEvent evt) {
            handleSharedLocationButtonSelected();
         }
      });
      setControl(comp);
   }

   private final ModifyListener fBasicModifyListener = evt -> updateLaunchConfigurationDialog();

   private void handleSharedRadioButtonSelected() {
      setSharedEnabled(isShared());
      updateLaunchConfigurationDialog();
   }

   private void setSharedEnabled(boolean enable) {
      fCustomSrcLocationText.setEnabled(enable);
      fCustomSrcLocationButton.setEnabled(enable);
   }

   private boolean isShared() {
      return fCustomSrcRadioButton.getSelection();
   }

   private void handleSharedLocationButtonSelected() {
      String currentContainerString = fCustomSrcLocationText.getText();
      IContainer currentContainer = getContainer(currentContainerString);
      SharedLocationSelectionDialog dialog = new SharedLocationSelectionDialog(getShell(), currentContainer, false, EMPTY_STRING);
      dialog.showClosedProjects(false);
      dialog.open();
      Object[] results = dialog.getResult();
      if ((results != null) && (results.length > 0) && (results[0] instanceof IPath)) {
         IPath path = (IPath) results[0];
         String containerName = path.toOSString();
         fCustomSrcLocationText.setText(containerName);
      }
   }

   private IContainer getContainer(String path) {
      Path containerPath = new Path(path);
      return (IContainer) ResourcesPlugin.getWorkspace().getRoot().findMember(containerPath);
   }

   class SharedLocationSelectionDialog extends ContainerSelectionDialog {

      private final String SETTINGS_ID = IDebugUIConstants.PLUGIN_ID + ".SHARED_LAUNCH_CONFIGURATON_DIALOG";

      public SharedLocationSelectionDialog(Shell parentShell, IContainer initialRoot, boolean allowNewContainerName, String message) {
         super(parentShell, initialRoot, allowNewContainerName, message);
      }

      @Override
      protected IDialogSettings getDialogBoundsSettings() {
         IDialogSettings settings = DebugUIPlugin.getDefault().getDialogSettings();
         IDialogSettings section = settings.getSection(SETTINGS_ID);
         if (section == null) {
            section = settings.addNewSection(SETTINGS_ID);
         }
         return section;
      }
   }

   @Override
   public String getName() {
      return LaunchConfigurationsMessages.getString("CustomisedLaunchConfigTab.LookupPath");
   }

   @Override
   public void initializeFrom(ILaunchConfiguration configuration) {
      try {
         boolean flag = configuration.getAttribute(USE_CUSTOM_SRC_PATH, false);
         if (flag) {
            fCustomSrcRadioButton.setSelection(true);
            setSharedEnabled(true);
            fCustomSrcLocationText.setText(configuration.getAttribute(CUSTOM_SRC_PATH, EMPTY_STRING));
         } else {
            fLocalRadioButton.setSelection(true);
            setSharedEnabled(false);
         }
      } catch (CoreException ce) {
         TestFrameworkPlugin.getDefault().getLog().log(ce.getStatus());
      }
   }

   @Override
   public void performApply(ILaunchConfigurationWorkingCopy configuration) {
      configuration.setAttribute(USE_CUSTOM_SRC_PATH, isShared());
      configuration.setAttribute(CUSTOM_SRC_PATH, fCustomSrcLocationText.getText());
   }

   @Override
   public void setDefaults(ILaunchConfigurationWorkingCopy configuration) {
      configuration.setAttribute(USE_CUSTOM_SRC_PATH, false);
   }

   @Override
   public boolean isValid(ILaunchConfiguration launchConfig) {
      if (isShared() && fCustomSrcLocationText.getText().equals(EMPTY_STRING)) {
         setErrorMessage(LaunchConfigurationsMessages.getString("CustomisedLaunchConfigTab.NoSourcePathSelected"));
         return false;
      }
      return true;
   }
}

File: CUTEProject/bundles/ch.hsr.ifs.testframework/src/ch/hsr/ifs/testframework/launch/CustomisedLaunchConfigTab.kt
/*******************************************************************************
 * Copyright (c) 2007-2011, IFS Institute for Software, HSR Rapperswil,
 * Switzerland, http://ifs.hsr.ch
 *
 * Permission to use, copy, and/or distribute this software for any
 * purpose without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 ******************************************************************************/
package ch.hsr.ifs.testframework.launch

import org.eclipse.cdt.launch.ui.CLaunchConfigurationTab
import org.eclipse.core.resources.IContainer
import org.eclipse.core.resources.ResourcesPlugin
import org.eclipse.core.runtime.CoreException
import org.eclipse.core.runtime.IPath
import org.eclipse.core.runtime.Path
import org.eclipse.debug.core.ILaunchConfiguration
import org.eclipse.debug.core.ILaunchConfigurationWorkingCopy
import org.eclipse.debug.internal.ui.DebugUIPlugin
import org.eclipse.debug.internal.ui.SWTFactory
import org.eclipse.debug.ui.IDebugUIConstants
import org.eclipse.jface.dialogs.IDialogSettings
import org.eclipse.swt.SWT
import org.eclipse.swt.events.ModifyListener
import org.eclipse.swt.events.SelectionAdapter
import org.eclipse.swt.events.SelectionEvent
import org.eclipse.swt.graphics.Font
import org.eclipse.swt.layout.GridData
import org.eclipse.swt.layout.GridLayout
import org.eclipse.swt.widgets.Button
import org.eclipse.swt.widgets.Composite
import org.eclipse.swt.widgets.Group
import org.eclipse.swt.widgets.Label
import org.eclipse.swt.widgets.Shell
import org.eclipse.swt.widgets.Text
import org.eclipse.ui.dialogs.ContainerSelectionDialog

import ch.hsr.ifs.testframework.TestFrameworkPlugin
import ch.hsr.ifs.testframework.model.Model


private const val EMPTY_STRING = ""
public const val CUSTOM_SRC_PATH = "customSrcPath"
public const val USE_CUSTOM_SRC_PATH = "useCustomSrcPath"

/**
 * @since 3.0
 */
//@SuppressWarnings("restriction")
class CustomisedLaunchConfigTab : CLaunchConfigurationTab() {

   protected lateinit var descriptionLabel: Label

   private lateinit var fLocalRadioButton: Button
   private lateinit var fCustomSrcRadioButton: Button
   private lateinit var fCustomSrcLocationText: Text
   private lateinit var fCustomSrcLocationButton: Button

   protected val cm = TestFrameworkPlugin.getModel()

   @Override
   override fun createControl(parent: Composite) {
      val font = parent.getFont()
      val comp = Composite(parent, SWT.NONE)
      createVerticalSpacer(comp, 1)

      val layout = GridLayout(1, true)
      layout.numColumns = 2
      layout.marginHeight = 0
      layout.marginWidth = 5
      comp.setLayout(layout)
      comp.setFont(font)

      val text = LaunchConfigurationsMessages["CustomisedLaunchConfigTab.SourceFolderSelection"]
      val group = SWTFactory.createGroup(comp, text, 3, 2, GridData.FILL_HORIZONTAL)
      val n_comp = SWTFactory.createComposite(group, parent.getFont(), 3, 3, GridData.FILL_BOTH, 0, 0)
      descriptionLabel = Label(n_comp, SWT.NONE)
      descriptionLabel.setText(LaunchConfigurationsMessages["CustomisedLaunchConfigTab.SourceDescText"])
      GridData().let{
         it.horizontalSpan = 3
         descriptionLabel.layoutData = it
      }

      fLocalRadioButton = createRadioButton(n_comp, LaunchConfigurationsMessages["CustomisedLaunchConfigTab.Default"])
      GridData().let{
         it.horizontalSpan = 3
         fLocalRadioButton.layoutData = it
      }

      fCustomSrcRadioButton = createRadioButton(n_comp, LaunchConfigurationsMessages["CustomisedLaunchConfigTab.CustomPath"])
      fCustomSrcRadioButton.addSelectionListener(object : SelectionAdapter() {
         override fun widgetSelected(evt: SelectionEvent) =
            handleSharedRadioButtonSelected()
      })

      fCustomSrcLocationText = SWTFactory.createSingleText(n_comp, 1)
      fCustomSrcLocationText.addModifyListener{ updateLaunchConfigurationDialog() }
      fCustomSrcLocationButton = createPushButton(n_comp, EMPTY_STRING, null)
      fCustomSrcLocationButton.addSelectionListener(object : SelectionAdapter() {
         override fun widgetSelected(evt: SelectionEvent) =
            handleSharedLocationButtonSelected()
      })
      setControl(comp)
   }

   private fun handleSharedRadioButtonSelected() {
      setSharedEnabled(isShared())
      updateLaunchConfigurationDialog()
   }

   private fun setSharedEnabled(enable: Boolean) {
      fCustomSrcLocationText.setEnabled(enable)
      fCustomSrcLocationButton.setEnabled(enable)
   }

   private fun isShared() =
         fCustomSrcRadioButton.getSelection()

   private fun handleSharedLocationButtonSelected() {
      val currentContainerString = fCustomSrcLocationText.getText()
      val currentContainer = getContainer(currentContainerString)
      val dialog = SharedLocationSelectionDialog(getShell(), currentContainer, false, EMPTY_STRING)
      dialog.showClosedProjects(false)
      dialog.open()
      val results = dialog.getResult()
      results?.let{
         if(it.size > 0 && it[0] is IPath) {
            fCustomSrcLocationText.text = (it[0] as IPath).toOSString()
         }
      }
   }

   private fun getContainer(path: String) =
         Path(path).let{
            ResourcesPlugin.getWorkspace().getRoot().findMember(it) as IContainer
         }

   private class SharedLocationSelectionDialog(parent: Shell, root: IContainer, allowNewContainerName: Boolean, message:String) : ContainerSelectionDialog(parent, root, allowNewContainerName, message) {

      private val SETTINGS_ID = IDebugUIConstants.PLUGIN_ID + ".SHARED_LAUNCH_CONFIGURATON_DIALOG"

      override protected fun getDialogBoundsSettings() = 
            DebugUIPlugin.getDefault().getDialogSettings().let{
               it.getSection(SETTINGS_ID) ?: it.addNewSection(SETTINGS_ID)
            }
   }

   override fun getName() = LaunchConfigurationsMessages["CustomisedLaunchConfigTab.LookupPath"]

   override fun initializeFrom(configuration: ILaunchConfiguration) {
      try {
         val flag = configuration.getAttribute(USE_CUSTOM_SRC_PATH, false)
         if (flag) {
            fCustomSrcRadioButton.selection = true
            setSharedEnabled(true)
            fCustomSrcLocationText.text = configuration.getAttribute(CUSTOM_SRC_PATH, EMPTY_STRING)
         } else {
            fLocalRadioButton.selection = true
            setSharedEnabled(false)
         }
      } catch (ce: CoreException) {
         TestFrameworkPlugin.getDefault().getLog().log(ce.getStatus())
      }
   }

   override fun performApply(configuration: ILaunchConfigurationWorkingCopy) {
      configuration.setAttribute(USE_CUSTOM_SRC_PATH, isShared())
      configuration.setAttribute(CUSTOM_SRC_PATH, fCustomSrcLocationText.getText())
   }

   override fun setDefaults(configuration: ILaunchConfigurationWorkingCopy) =
         configuration.setAttribute(USE_CUSTOM_SRC_PATH, false)

   override fun isValid(launchConfig: ILaunchConfiguration) =
      if (isShared() && fCustomSrcLocationText.getText().equals(EMPTY_STRING)) {
         setErrorMessage(LaunchConfigurationsMessages["CustomisedLaunchConfigTab.NoSourcePathSelected"])
         false
      } else {
         true
      }
}